File: Models/random_forest.R
library(ggplot2)
library(data.table)
library(caret)
library(scales)
library(randomForest)
library(SDMTools)
library(pROC)
library(klaR)
library(xtable)

set.seed(0)

#----------- Part 1: Read data, split out validation set and split out labels
train <- read.csv('~/Desktop/CSML/bioinformatics/coursework/data/train.csv')
test <- read.csv('~/Desktop/CSML/bioinformatics/coursework/data/test.csv')
blind_test <- read.csv('~/Desktop/CSML/bioinformatics/coursework/data/blind_test.csv')

# # Use important features only
# load(file = "~/Desktop/CSML/bioinformatics/coursework/important_features.RData")
# train <- train[, important_features]
# test <- test[, important_features]

smp_size <- floor(0.75 * nrow(train))
train_ind <- sample(seq_len(nrow(train)), size = smp_size)
train_set <- train[train_ind, ]
val_set <- train[-train_ind, ]

y_train <- factor(train_set$class, level=c('cyto', 'mito', 'nuclear', 'secreted'))
names_train <- train_set$name
X_train <- subset(train_set, select=-c(name, class, sequence))

y_val <- factor(val_set$class, level=c('cyto', 'mito', 'nuclear', 'secreted'))
names_val <- val_set$name
X_val <- subset(val_set, select=-c(name, class, sequence))

y_test <- factor(test$class, level=c('cyto', 'mito', 'nuclear', 'secreted'))
names_test <- test$name
X_test <- subset(test, select=-c(name, class, sequence))


#-------- Part 2: Fit Random Forest model
train_set <- X_train
train_set$y <- y_train
rf_model <- randomForest(y ~ ., data=train_set, ntree = 1000, mtry = 20, nodesize=1)

# Train and validation set predictions
train_pred <- predict(rf_model)
print("Training Accuracy:")
print(sum(train_pred==y_train)/length(y_train))
table(y_train, train_pred)

val_pred <- predict(rf_model, newdata=X_val)
print("Validation Accuracy:")
print(sum(val_pred==y_val)/length(y_val))
table(y_val, val_pred)
print(xtable(data.frame(table(y_val, val_pred))), include.rownames=FALSE) # latex table format

# Feature importance
importance <- data.frame(rf_model$importance)
importance$Feature <- rownames(importance)
importance <- data.table(importance)[order(-MeanDecreaseGini), ]
setcolorder(importance, c('Feature', 'MeanDecreaseGini'))
importance <- importance[MeanDecreaseGini>=46]  # median value
importance$Feature <- factor(importance$Feature)
importance$Feature <- reorder(importance$Feature, -importance$MeanDecreaseGini)

# Plot and save as PNG file
g <- ggplot(data=importance, aes(x=Feature, y=MeanDecreaseGini, fill=Feature)) + geom_bar(stat='identity')
g <- g + labs(x='Feature', y='Mean Decrease Gini Index') + coord_flip() + guides(fill=FALSE)
g + scale_x_discrete(limits = rev(levels(importance$Feature)))

png(filename="~/Desktop/CSML/bioinformatics/coursework/figures/feature_importance.pdf")
g + scale_x_discrete(limits = rev(levels(importance$Feature)))
dev.off()

# Save importance features for later use
important_features <- as.character(importance$Feature)
important_features <- c(important_features, 'name', 'sequence', 'class')
save(important_features, file = "~/Desktop/CSML/bioinformatics/coursework/important_features.RData")


#------ Part 3: Assess performance on test set

# Re-train on full training set
train <- read.csv('~/Desktop/CSML/bioinformatics/coursework/data/train.csv')
train <- train[, important_features]  # important features only
y_train <- factor(train$class, level=c('cyto', 'mito', 'nuclear', 'secreted'))
X_train <- subset(train, select=-c(name, class, sequence))
train_set <- X_train
train_set$y <- y_train
rf_model <- randomForest(y ~ ., data=train_set, ntree = 1000, mtry = 20, nodesize=1)
train_pred <- predict(rf_model)

# Test set accuracy
test_pred <- predict(rf_model, newdata=X_test)
print("Test Accuracy:")
print(sum(test_pred==y_test)/length(y_test))

# Confusion matrix
table(y_test, test_pred)
xtable(data.frame(table(y_test, test_pred))) # latex table format

# Plot ROC curves and compute AUC
test_pred_prob <- data.table(data.frame(predict(rf_model, newdata=X_test, type='prob')))
test_pred_prob$prediction <- test_pred
test_pred_prob$true_label <- y_test

class_prob_all <- data.frame()
auc_scores <- c()

for (class_name in c('cyto', 'mito', 'nuclear', 'secreted')) {
  class_prob <- subset(test_pred_prob, select=c(eval(as.symbol(class_name)), prediction, true_label))[order(-eval(as.symbol(class_name)))]
  print(names(class_prob))
  
  # ROC curve and calculate AUC score
  class_prob$FP <- 0; class_prob$TP <- 0
  class_prob[class_prob$prediction == class_name & class_prob$true_label == class_name, ]$TP <- 1
  class_prob[class_prob$prediction == class_name & class_prob$true_label != class_name, ]$FP <- 1
  class_prob$pos <- 0; class_prob$neg <- 0
  class_prob[class_prob$true_label == class_name, ]$pos <- 1
  class_prob[class_prob$true_label != class_name, ]$neg <- 1
  class_prob$TPR <- cumsum(class_prob$TP) / sum(class_prob$TP)
  class_prob$FPR <- cumsum(class_prob$FP) /sum(class_prob$FP)
  class_prob <- subset(class_prob, select=c(TPR, FPR))
  n <- nrow(class_prob)
  
  area <- sum(((class_prob$TPR[-1]+class_prob$TPR[-n])/2) * (class_prob$FPR[-1]-class_prob$FPR[-n]))
  auc_scores <- c(auc_scores, area)
  class_prob$class <- class_name
  class_prob_all <- rbind(class_prob_all, class_prob)
}

class_prob_all$class <- factor(class_prob_all$class)
g <- ggplot(data=class_prob_all, aes(x=FPR, y=TPR, color=class)) + geom_line(size=0.8)
g <- g + labs(x='False positive rate', y='True positive rate', color='')
g + theme(legend.position="top") + geom_abline(intercept=0, slope=1, linetype = 3) + ylim(c(0, 1)) + xlim(c(0, 1))

png(filename="~/Desktop/CSML/bioinformatics/coursework/figures/ROC_curve.png")
g + theme(legend.position="top")
dev.off()

# AUC table and mean AUC
print("Mean AUC:")
(sum(test$class=='cyto')*auc_scores[1] + sum(test$class=='mito')*auc_scores[2] + sum(test$class=='nuclear')*auc_scores[2] +
  sum(test$class=='secreted')*auc_scores[4]) /nrow(test)
auc_scores <- data.frame(Class=c('cyto', 'mito', 'nuclear', 'secreted'), AUC_score=auc_scores)
xtable(auc_scores)  # latex output


##----- Part 4: Predictions on blind test set
protein_names <- blind_test$name
blind_test <- subset(blind_test, select=-c(name, sequence))
blind_test_pred <- data.table(data.frame(predict(rf_model, newdata=blind_test, type='prob')))
blind_test_pred$name <- protein_names
blind_test_pred <- melt(blind_test_pred, id.vars='name')
blind_test_pred <- blind_test_pred[, .SD[value==max(value)], by=name]
blind_test_pred <- blind_test_pred[order(name)]
names(blind_test_pred) <- c('Sequence Identifier', 'Class Prediction', 'Value')
xtable(blind_test_pred)

File: Models/random_forest.py
from __future__ import division
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split
from sklearn.datasets import make_classification
from sklearn.ensemble import ExtraTreesClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import confusion_matrix

def accuracy(pred, actual):
    """Returns percentage of correctly classified labels"""
    return sum(pred==actual) / len(actual)

def main():

	# Import data and split out labels
	train = pd.read_csv('../data/train.csv')
	test = pd.read_csv('../data/test.csv')

	# Split into labels, names and data
	y_train = train['class']
	names_train = train['name']
	X_train = train.drop(['class', 'name', 'sequence'], axis=1)

	y_test = test['class']
	names_train = test['name']
	X_test = test.drop(['class', 'name', 'sequence'], axis=1)

	# Optimize a random forest model using grid search
	rf = RandomForestClassifier()

	param_grid = {
	    'n_estimators': [500], 
	    'max_depth': [15, 20, 25],
	    'max_features': [8, 10, 12],
	    'n_jobs': [30],
	    'min_samples_leaf': [1, 2, 3]
	}

	grid_rf = GridSearchCV(rf, param_grid, cv=2, verbose=3)
	grid_rf.fit(X_train, y_train)
	print("#-------- DONE WITH GRID SEARCH.")
	best_model = grid_rf.best_estimator_
	best_params = grid_rf.best_params_ 
	scores = grid_rf.grid_scores_
	print(best_params)

	# Calculate cross validation accuracy
	rf = RandomForestClassifier(n_estimators=500, criterion='entropy', max_features=8, max_depth=15,
	                            min_samples_leaf = 1, bootstrap=True, oob_score=True, n_jobs=30, random_state=0)
	print(np.mean(cross_val_score(rf, X_train, y_train, cv=5)))

	# # Fit to full training data and table feature importances
	# rf = rf.fit(X_train, y_train)
	# importances = rf.feature_importances_
	# importance = pd.DataFrame(importances, index=X_train.columns, columns=["importance"])
	# # importance.sort('importance', ascending=0)

	# # Print train and test accuracy
	# y_train_pred = rf.predict(X_train)
	# y_test_pred = rf.predict(X_test)
	# print("Training Accuracy = %f" % accuracy(y_train_pred, y_train))
	# print("Test Accuracy = %f" % accuracy(rf.predict(X_test), y_test))

	# confusion_matrix(np.array(y_test), np.array(y_test_pred))

if __name__ == "__main__":
	main()